### ISSUE TITLE
FST codec on Topic (PUB\SUB).

### ISSUE THREAD

**Issue by aam1go** [redisson/redisson#3411]
So i suggest. There is a way to do FST codec on Topic (PUB\SUB). When i tried to use it. It didnt work. Then i though that im choose a wrong way. And then i Make log of redisson to debug "logging.level.org.redisson=debug". 
So what i see when i use publish method : 

> org.redisson.command.RedisExecutor       : connection released for command (PUBLISH) and params [SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 48, cap: 256)] from slot NodeSource [slot=0, addr=null, redisClient=null, redirect=null, entry=null] using connection RedisConnection@1597660839 [redisClient=[addr=redis://db.test.me:6379], channel=[id: 0xc125e7d4, L:/192.168.126.1:59815 - R:db.test.me/192.168.126.142:6379], currentCommand=CommandData [promise=RedissonPromise [promise=ImmediateEventExecutor$ImmediatePromise@75ecf0f1(success: 1)], command=(PUBLISH), params=[SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 48, cap: 256)], codec=**org.redisson.client.codec.StringCodec**]]

As we see there is using still StringCodec.  **Then i tried to use FST codec not for PUB\SUB. Just for save info to redisson. And then it works**. My configuration of FST is simple :

>         FSTConfiguration fstConfiguration = FSTConfiguration.createDefaultConfiguration();
>         fstConfiguration.setForceSerializable(true);

There is a way to use FSTCodec for Pub\Sub, not only for store data? Version of redisson 3.14

**Comment by mrniko** [comment]
Try to use MarshallingCodec, because FSTCodec throws NPE in some cases.

**Comment by aam1go** [comment]
> Use MarshallingCodec, because FSTCodec throws NPE in some cases.

Thank you. Is MarshallingCodec fastest then FST or closely fast? Where i can watch top codecs by speed?

**Comment by aam1go** [comment]
And. Then i using MarshallingCodec. Make Data class implements Serializable to test this feature. Publish data message.

> org.redisson.command.RedisExecutor       : connection released for command (PUBLISH) and params [SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 102, cap: 256)] from slot NodeSource [slot=0, addr=null, redisClient=null, redirect=null, entry=null] using connection RedisConnection@1208428384 [redisClient=[addr=redis://db.test.me:6379], channel=[id: 0x1e90b288, L:/192.168.126.1:55474 - R:db.test.me/192.168.126.142:6379], currentCommand=CommandData [promise=RedissonPromise [promise=ImmediateEventExecutor$ImmediatePromise@4487bdf5(success: 1)], command=(PUBLISH), params=[SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 102, cap: 256)], **codec=org.redisson.client.codec.StringCodec**]]

Still StringCodec for PUB\SUB. I think this not change. I used jsonJackson. No changes. @mrniko

**Comment by mrniko** [comment]
both codecs encode string in binary format.

**Comment by aam1go** [comment]
yep, i though too. But there is next problem then codec for pub\sub not JsonJacksonCodec. 

>     @Scheduled(cron = "0/30 * * * * ?")
>     public void testPub() {
>         log.info("pub start");
>         OnSimpleEvent pubEvent = new OnSimpleEvent(1, 1, "Test");
>         servicesTopic.publish(pubEvent);
>     }
> 
>     private void testServiceSimplestTopic() {
>         servicesTopic.addListener(OnSimpleEvent.class, (charSequence, onSimpleEvent) -> {
>             log.info("OnSimpleEvent {}", onSimpleEvent);
>         });
>     }
> 

This code shows only: 

> 2021-02-09 17:00:50.002  INFO 15684 --- [   scheduling-1] d.k.web.workers.ServicesEventListener    : pub start
> 2021-02-09 17:00:50.002 DEBUG 15684 --- [   scheduling-1] org.redisson.command.RedisExecutor       : acquired connection for command (PUBLISH) and params [SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 101, cap: 256)] from slot NodeSource [slot=0, addr=null, redisClient=null, redirect=null, entry=null] using node db.test.me/192.168.126.142:6379... RedisConnection@1108733459 [redisClient=[addr=redis://db.test.me:6379], channel=[id: 0xc17219ec, L:/192.168.126.1:52087 - R:db.test.me/192.168.126.142:6379], currentCommand=null]
> 2021-02-09 17:00:50.003 DEBUG 15684 --- [sson-netty-2-11] org.redisson.command.RedisExecutor       : connection released for command (PUBLISH) and params [SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 101, cap: 256)] from slot NodeSource [slot=0, addr=null, redisClient=null, redirect=null, entry=null] using connection RedisConnection@1108733459 [redisClient=[addr=redis://db.test.me:6379], channel=[id: 0xc17219ec, L:/192.168.126.1:52087 - R:db.test.me/192.168.126.142:6379], currentCommand=CommandData [promise=RedissonPromise [promise=ImmediateEventExecutor$ImmediatePromise@5dc14ea4(success: 1)], command=(PUBLISH), params=[SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 101, cap: 256)], codec=org.redisson.client.codec.StringCodec]]

String codec is Ok if both codecs using string in binary format. BUT listener not works. If i use JsonJackson codec in pub sub is ok: 

> 2021-02-09 17:02:35.001  INFO 18884 --- [   scheduling-1] d.k.web.workers.ServicesEventListener    : pub start
> 2021-02-09 17:02:35.006 DEBUG 18884 --- [   scheduling-1] org.redisson.command.RedisExecutor       : acquired connection for command (PUBLISH) and params [SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 86, cap: 256)] from slot NodeSource [slot=0, addr=null, redisClient=null, redirect=null, entry=null] using node db.test.me/192.168.126.142:6379... RedisConnection@1329871175 [redisClient=[addr=redis://db.test.me:6379], channel=[id: 0x98396e22, L:/192.168.126.1:52284 - R:db.test.me/192.168.126.142:6379], currentCommand=null]
> 2021-02-09 17:02:35.007 DEBUG 18884 --- [isson-netty-2-9] org.redisson.command.RedisExecutor       : connection released for command (PUBLISH) and params [SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 86, cap: 256)] from slot NodeSource [slot=0, addr=null, redisClient=null, redirect=null, entry=null] using connection RedisConnection@1329871175 [redisClient=[addr=redis://db.test.me:6379], channel=[id: 0x98396e22, L:/192.168.126.1:52284 - R:db.test.me/192.168.126.142:6379], currentCommand=CommandData [promise=RedissonPromise [promise=ImmediateEventExecutor$ImmediatePromise@45e3ff21(success: 1)], command=(PUBLISH), params=[SERVICES, PooledUnsafeDirectByteBuf(ridx: 0, widx: 86, cap: 256)], codec=org.redisson.client.codec.StringCodec]]
> **2021-02-09 17:02:35.009  INFO 18884 --- [   redisson-3-2] d.k.web.workers.ServicesEventListener    : OnSimpleEvent OnSimpleEvent(MaxSize=1, carosta=1, mish=Test)**

AS WE SEE UPPER In case if JsonJacksonCodec we still see StringCodec by Listener is working. I dont know why it is so. But i need to make my PUB\SUB resolution fastest. But see this issue :(. So even i dont use manually codec. Comment setCodec at configuration and topic(Name) (without) codec I.E default codec: Event listener still NOT WORKING. 
This works only with JsonJackson codec manually. @mrniko

**Comment by mrniko** [comment]
Could you provide test code to reproduce the issue?

**Comment by aam1go** [comment]
> Could you provide test code to reproduce the issue?

[Sure.](https://dropmefiles.com/wgbwc)
Maven project Simple. At first for be sure that i made do i create this project by your question. And for me. I tried to use only 1-2 classes without Spring. So it works. With default codec is Marshalling. I tried to do same in Spring boot application. But is doesnt work. At this folder, that i download StartRedisTest and StartRedisTestBoot with simple logging to see that message by topic deliver or not.

**Comment by mrniko** [comment]
The problem is in classloader.

type.getClassLoader() = {RestartClassLoader@5632} 
message.getClass().getClassLoader() = {ClassLoaders$AppClassLoader@5626}

**Comment by mrniko** [comment]
To resolve this issue define classloader in codec constructor:

```java
config.setCodec(new MarshallingCodec(getClass().getClassLoader()));
```

**Comment by aam1go** [comment]
> To resolve this issue define classloader in codec constructor:
> 
> ```java
> config.setCodec(new MarshallingCodec(getClass().getClassLoader()));
> ```

Ok i fix this. This question is resolve. Thank You!)
There is some question was yesterday but then it wasnt priority. But now its important for me.

So i want to increase performance by choose best stable Codec. So you advice Marshalling codec instead of FSTCodec. So i need to know how to use Marshaling codec without Serializable data class. In FST there is a feature  

>     FSTConfiguration fstConfiguration = FSTConfiguration.createDefaultConfiguration();
>     fstConfiguration.setForceSerializable(true);

May i do same in another codecs? i dont see this in Marshalling codec. Ty

**Comment by mrniko** [comment]
> May i do same in another codecs

You can use JsonJacksonCodec
